jaeger-core/src/main/java/io/jaegertracing/internal/reporters/FilteringReporter.java

```diff
@@ -0,0 +1,116 @@
+import io.jaegertracing.spi.Reporter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.LongAdder;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * FilteringReporter drops spans below a specific threshold and buffers the remaining spans in memory and
+ * sends them to a delegate only once a parent span exceeds a specified threshold.
+ */
+@ToString
+@Slf4j
+public class FilteringReporter implements Reporter {
+  public static final long DEFAULT_FILTER_SPANS_UNDER_MICROS = 0L;
+  public static final long DEFAULT_DEFER_SPANS_UNDER_MICROS = 0L;
+
+  private final Reporter delegate;
+  private final long filterSpansUnderMicros;
+  private final long deferSpansUnderMicros;
+  private final Metrics metrics;
+
+  private final Map<Long, List<JaegerSpan>> pendingByParent = new ConcurrentHashMap<>();
+  private final LongAdder pendingCount = new LongAdder();
+
+  public FilteringReporter(
+      Reporter delegate, long filterSpansUnderMicros, long deferSpansUnderMicros, Metrics metrics) {
+    this.delegate = delegate;
+    this.filterSpansUnderMicros = filterSpansUnderMicros;
+    this.deferSpansUnderMicros = deferSpansUnderMicros;
+    this.metrics = metrics;
+  }
+
+  @Override
+  public void report(JaegerSpan span) {
+    final JaegerSpanContext context = span.context();
+    final List<JaegerSpan> pendingChildren = pendingByParent.remove(context.getSpanId());
+
+    if (span.getDuration() < filterSpansUnderMicros) {
+      metrics.filteredSpans.inc(1);
+      return;
+    }
+
+    if (span.getDuration() < deferSpansUnderMicros) {
+      defer(span, context, pendingChildren);
+    } else {
+      // report pending if any, then this span
+      if (pendingChildren != null) {
+        pendingChildren.forEach(delegate::report);
+        final int count = pendingChildren.size();
+        pendingCount.add(-1 * count);
+        metrics.deferredSpansPending.update(pendingCount.longValue());
+        metrics.deferredSpansSent.inc(count);
+      }
+      delegate.report(span);
+    }
+  }
+
+  private void defer(final JaegerSpan span, final JaegerSpanContext context,
+      final List<JaegerSpan> pendingChildren) {
+    final long parentId = context.getParentId();
+    if (parentId != 0) {
+      pendingCount.increment();
+      metrics.deferredSpansPending.update(pendingCount.longValue());
+      metrics.deferredSpansStarted.inc(1);
+      pendingByParent.compute(parentId, (id, spans) -> {
+        if (spans == null) {
+          if (pendingChildren == null) {
+            spans = new ArrayList<>(1);
+          } else {
+            spans = pendingChildren;
+          }
+        } else if (pendingChildren != null) {
+          spans.addAll(pendingChildren);
+        }
+        spans.add(span);
+        return spans;
+      });
+    } else if (pendingChildren != null) {
+      final int count = pendingChildren.size();
```

Comment:
The fact that we're counting `span` here by not subtracting one should be commented here so someone doesn't "fix" this later.